### ISSUE TITLE
[BUG] Using a tag in an exit lock doesn't work if there is a tag category

### ISSUE THREAD

**Issue by RianeNelson** [evennia/evennia#2139]
#### Describe the bug
If you try to use an exit with a tag for the lock, you can't have a category for the tag.
![2020-05-18 (1)](https://user-images.githubusercontent.com/42573189/82182853-2d476900-98a2-11ea-8074-030cf9819ed3.png)
![2020-05-18](https://user-images.githubusercontent.com/42573189/82182856-2e789600-98a2-11ea-9444-d0a71dd0002e.png)

#### To Reproduce
1. Create an exit with traverse:tag(tagname)
2. Create a tag (tagname) with a category (tagcategory) on your character
3. Try to go through the exit. Won't allow it.
4. Remove the tag category from the character so it's just the tag name and it works.

#### Expected behavior
I would expect you should be able to use the exit regardless if there is a category?
#### Environment, Evennia version, OS etc
Evennia 0.9.0 (rev 10a6685f) (rev 10a6685f)
    OS: nt
    Python: 3.8.2
    Twisted: 19.10.0
    Django: 2.2.12
#### Additional context

I have also tried to keep the category on the object tag, and then add it to the lock to see if that works, but it doesn't appear to. Once you have the category, it appears to cease working.

**Comment by RianeNelson** [comment]
So here's another thing I think may be a bug
if you remove the lock for tags, even if it doesn't show it in the exit (when you do examine exit, or when you pull up the file on the django admin site) it still acts like it's there.
And the only way around it that I can find is deleting the exit entirely and remaking it
it acts like the broken lock is still there and then the only way to get by the exit is unquelling.

**Comment by RianeNelson** [comment]
I just tried something. If I make a brand new exit, and add the lock without using the exit yet, then add a tag without a category, it lets me through ONE time.
But after that I can't go through.
I can go through once, if it's a new exit, and has not yet been used.

**Comment by Griatch** [comment]
I cannot test right now, but traversing a (vanilla) exit should not modify it in any way and there could be no state change to make the lock invalid on a second try. A very strange effect.

**Comment by Griatch** [comment]
## Original bug report 
To be clear, putting  `traverse:tag(tagname)` an exit means this: 

- Allow traversal of this exit if and only if: 
  - The traversing object has the `tagname` tag ...
  - ... and that tag has no category (category == `None`)

Tags are unique in `key + category` pairs. One such combination is `tagname + None`, which is _not_ the same as `tagname + tagcategory`. So if you putting a tag `tagname` with a tag-category of `tagcategory` on yourself you are failing the second condition of this check and will not pass the lock. You have simply created a different tag than your lock is checking for. This appears to behave as expected.

If you want the specific lock to honor your specific tag, it sounds like you need to put a lock `traverse:tag(tagname,tagcategory)` on it. 

So this doesn't appear to be a bug. I have expanded the Tags docs a little to make this use of categories [a little clearer](https://github.com/evennia/evennia/wiki/Tags#properties-of-tags-and-aliases-and-permissions). 

## Going through the lock once

This is probably a genuine bug; I suspect your first attempt will fail too if:

 - you set the lock in-game (rather than in the web admin) or 
 - if you reload the server before going through the first time. 

If I'm right, the reason you can go through on the first attempt and not on the second would then be because of caches not getting updated in contents when you set the lock in the admin - so the first attempt passes because the system has not seen the lockstring change yet. That inconsistency should be fixed.

**Comment by RianeNelson** [comment]
Using a category at all doesn't work either. If I have a tag with a category on an object and do traverse:tag(tagname, tagcategory) it doesn't allow the object to pass. The same issue exists with the object 'holds' lock, as well.

**Comment by Griatch** [comment]
@RianeNelson I cannot replicate this. This is what I tried: 
```python
obj.tags.add("test1", "category1")
exit.locks.add("traverse:tag(test1,category1)")
exit.access(obj, "traverse")   # this returns True, so we pass the 'traverse' lock
```
So this passes correctly for me. Could you check if you see your problem also when testing this way?

**Comment by RianeNelson** [comment]
@Griatch it looks like when I recreated the exit that does work, my bad! If you add the category in the lock when you create the lock, there's no issue. It seems to just break the exit when you don't add the category.

**Comment by RianeNelson** [comment]
@Griatch I had been adding tags via the web admin because I couldn't get it to work using in-game commands. It keeps showing me the previous working tag and an error with it, even when I'm using a different tag: 
![2020-05-18 (9)](https://user-images.githubusercontent.com/42573189/82251891-46383480-990b-11ea-8541-6200f0414893.png)

**Comment by strikaco** [comment]
Without addressing the synchronization issue between the admin and CLI (transaction rollback?), might the root issue here have been resolved by #2149 ? Is this still a problem?

Before ~20 days ago, the lock checker was ignoring categories when evaluating tags.

**Comment by Griatch** [comment]
Yes, I suspect this was resolved as a side effect of #2149. Can't reproduce. I'll close this for now. @RianeNelson - comment here if you still have issues and I can open this again. :slightly_smiling_face:
